[SPARK-12559][SPARK SUBMIT] fix --packages for stand-alone cluster mode

skonto opened the pull request:
## What changes were proposed in this pull request?

Fixes --packages flag for the stand-alone case in cluster mode. Adds to the driver classpath the jars that are resolved via ivy along with any other jars passed to `spark.jars`. Jars not resolved by ivy are downloaded explicitly to a tmp folder on the driver node. Similar code is available in SparkSubmit so we refactored part of it to use it at the DriverWrapper class which is responsible for launching driver in standalone cluster mode.

Note: In stand-alone mode `spark.jars` contains the user jar so it can be fetched later on at the executor side.

## How was this patch tested?
Manually by submitting a driver in cluster mode within a standalone cluster and checking if dependencies were resolved at the driver side.

skonto committed:
fix --packages for standalone

skonto commented:
@vanzin pls review.

SparkQA commented:
**[Test build #79598 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79598/testReport)** for PR 18630 at commit [`81a5dee`](https://github.com/apache/spark/commit/81a5deeb2f8d152242f9248f8e5d5c0dd3e1f1ea).

SparkQA commented:
**[Test build #79598 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79598/testReport)** for PR 18630 at commit [`81a5dee`](https://github.com/apache/spark/commit/81a5deeb2f8d152242f9248f8e5d5c0dd3e1f1ea).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79598/
Test FAILed.

SparkQA commented:
**[Test build #79599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79599/testReport)** for PR 18630 at commit [`9816704`](https://github.com/apache/spark/commit/9816704abcbe12bf99e8e735bf8ede679dcd8cd1).

SparkQA commented:
**[Test build #79599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79599/testReport)** for PR 18630 at commit [`9816704`](https://github.com/apache/spark/commit/9816704abcbe12bf99e8e735bf8ede679dcd8cd1).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79599/
Test FAILed.

jerryshao commented:
Are you trying to support `--packages` in standalone cluster?

jerryshao commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -473,6 +474,12 @@ object SparkSubmit extends CommandLineUtils {
         sysProp = "spark.driver.extraLibraryPath"),
 
+      // Standalone only - propagate attributes for dependency resolution at the driver side
+      OptionAssigner(args.packages, STANDALONE, CLUSTER, sysProp = "spark.jars.packages"),
```
You can merge this with mesos cluster `OptionAssigner` like `OptionAssigner(args.packages, STANDALONE | MESOS, CLUSTER, sysProp = "spark.jars.packages")`.

skonto replied:
ok...

jerryshao commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +75,50 @@ object DriverWrapper {
+
+    // filter out the user jar
+    jars = jars.split(",").filterNot(_.contains(userJar.split("/").last)).mkString(",")
+    jars = Option(jars).map(SparkSubmit.downloadFileList(_, hadoopConf)).orNull
```
The API interface is changed, this is not compilable.

skonto replied:
Ok will fix that...

jerryshao commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +75,50 @@ object DriverWrapper {
+    var repositories = sys.props.get("spark.jars.repositories").orNull
+    val hadoopConf = new HadoopConfiguration()
+    val childClasspath = new ArrayBuffer[String]()
+    var jars = sys.props.get("spark.jars").orNull
```
Here I think `jars` and `files` can only be remote resources, otherwise how can remote driver visit SparkSubmit local resources? So I think some defensive code may be necessary.

skonto replied:
Ivy does the resolution and downloads them locally at the file system's node where the driver is launched on. So then you use the hadoop utilities to fetch them. That's the idea about jars. I was only focusing on packages... but yes need to write some more code to handle files and remote jars already passed though.

skonto commented:
@jerryshao 

> 

Are you trying to support --packages in standalone cluster?

Yes that is my intention as stated in the description. I will proceed with the changes thnx for the review.

skonto commented:
One thing I noticed is that if you use https you need to set : --conf spark.ssl.enabled=true --conf spark.ssl.protocol=TLS but this does nto apply to the user jar. I think that is an issue.

SparkQA commented:
**[Test build #79669 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79669/testReport)** for PR 18630 at commit [`1f92106`](https://github.com/apache/spark/commit/1f92106fe04bfc1fd3d03523dd4fe160eb91a634).

skonto commented:
@jerryshao @vanzin  I updated the PR, pls review. In standalone cluster mode with this new update I can resolve jars paths of the form: /tmp/tmp7255787727862122372/jarname.jar,
file:///home/ubuntu/.ivy2/jars/jarname.jar
There is no need to download spark.files at the driver side this is taken care at the executor side as long as the property is preserved.

SparkQA commented:
**[Test build #79669 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79669/testReport)** for PR 18630 at commit [`1f92106`](https://github.com/apache/spark/commit/1f92106fe04bfc1fd3d03523dd4fe160eb91a634).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79669/
Test FAILed.

skonto committed:
fix style

SparkQA commented:
**[Test build #79670 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79670/testReport)** for PR 18630 at commit [`b75387c`](https://github.com/apache/spark/commit/b75387cf4cf2debc04979f2416efa01f82680b3c).

SparkQA commented:
**[Test build #79670 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79670/testReport)** for PR 18630 at commit [`b75387c`](https://github.com/apache/spark/commit/b75387cf4cf2debc04979f2416efa01f82680b3c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79670/
Test PASSed.

jerryshao commented:
I'm wondering if we can prepare all the resources in `SparkSubmit` and upload to HDFS's staging folder, then driver could download them from HDFS and add to classpath. This is what basically YARN#client does, I think maybe for standalone cluster mode we could do this also, so that the resources preparation logics are still in `SparkSubmit`. Just my two cents.

skonto commented:
@jerryshao I agree I had shared this thought with @vanzin for utilizing the distributed cache or whatever is available there in the other PR for mesos. For mesos I am also going to discuss this with Mesosphere guys. Flink has one layer for all: https://issues.apache.org/jira/browse/FLINK-6177
Check my comment here: https://github.com/apache/spark/pull/18587

jerryshao commented:
One layer for different cluster managers seems promising, but looks like it requires lots of refactoring works (trying to build a Spark own distributed cache and change the existing way to distribute resources). Maybe a quick fix is enough (I don't think there's lots of users will use standalone cluster mode).

skonto commented:
I agree that's why I didnt try to do the refactoring... I thought it would take a long time... As for the cluster mode some people use it due to the supervise feature, guess not too many, but dont have a clear view to be honest. @vanzin @srowen thoughts?

vanzin commented:
As with the Mesos change, I'm ok with this approach until there's some sort of distributed cache available for those cluster managers. I still need to review the code, though.

@skonto please mention in the PR title that this is for cluster mode.

skonto commented:
@vanzin thnx done, I changed the title. I am willing to work on the distributed cache but needs time for sure.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -18,8 +18,17 @@
+import java.nio.file.Files
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
```
Pick a style; either use `mutable.Foo` everywhere, or import each class.

skonto replied:
Sure...

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +77,68 @@ object DriverWrapper {
+
+  // R or Python are not supported in cluster mode so download the jars to the driver side
+  private def setupDependencies(loader: MutableURLClassLoader, userJar: String): Unit = {
+    val packagesExclusions = sys.props.get("spark.jars.excludes").orNull
```
Hmm, this is too much copy & paste for my liking. Could you instead refactor this code into a helper class or object (e.g. `IvytUtils`)?

skonto replied:
Its not exactly the same with the code in SparkSubmit but I can give it a shot. I thought about too.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +77,68 @@ object DriverWrapper {
     }
   }
+
+  // R or Python are not supported in cluster mode so download the jars to the driver side
```
It's not clear what R and Python have to do with downloading jar files.

skonto replied:
Just saying we dont covering them.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -43,7 +52,7 @@ object DriverWrapper {
 
         val currentLoader = Thread.currentThread.getContextClassLoader
-        val userJarUrl = new File(userJar).toURI().toURL()
+        val userJarUrl = new File(userJar).toURI.toURL
```
No need for this change.

skonto replied:
Its more scala style that I prefer but anyway...

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +77,68 @@ object DriverWrapper {
+    if (jars != null) {childClasspath ++= jars.split(",")}
+
+    for (jar <- childClasspath) {
+        SparkSubmit.addJarToClasspath(jar, loader)
```
So now all the stuff in `spark.jars` that actually exists on the node will be automatically added to the driver's classpath, which is a change in behavior from before. It's probably ok, but please call out this kind of thing in the commit message.

skonto replied:
Ok I will do so...

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +77,68 @@ object DriverWrapper {
+    // Filter out the user jar
+    jars = jars.split(",").filterNot(_.contains(userJar.split("/").last)).mkString(",")
+    jars = Option(jars)
+      .map(SparkSubmit.downloadFileList(_, targetDir, sparkProperties, hadoopConf)).orNull
```
You're also implementing a separate feature that is not part of SPARK-12559. You should at the very least call this out in the commit message. I wouldn't be surprised if there's already a bug tracking this, though, so you should take a look.

skonto replied:
I cannot omit the jars here... I skipped them earlier in SparkSubmit so I have to do this somewhere.. I am just migrating the place where this action takes place...

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +77,68 @@ object DriverWrapper {
+      .map {pName => sys.props.get(pName)
+        .map{pValue => sparkProperties.put(pName, pValue)}}
+
+    jars = Option(jars).map(SparkSubmit.resolveGlobPaths(_, hadoopConf)).orNull
```
These helper methods in `SparkSubmit` should probably be moved to a more properly-named object if they're going to be used in other places. e.g., instead of my previous `IvyUtils` suggestion, you could have a `DependencyUtils` or something like that having all these methods.

Then you could use that method in `DriverRunner` instead of its own `downloadUserJar`.

vanzin reviewed (commented):
Please write a more descriptive commit message. There's a lot going on here that gets no mention there.

I also dislike that there's a bunch of code that is being duplicated now; some re-factoring is in order here.

SparkQA commented:
**[Test build #79906 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79906/testReport)** for PR 18630 at commit [`c559a08`](https://github.com/apache/spark/commit/c559a08028633394371e784930dda0f8ff3f19ea).

skonto commented:
@vanzin I did the required refactoring, addressing the issues you mentioned and added to the description a better message so its available in git log when this PR is merged.

SparkQA commented:
**[Test build #79906 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79906/testReport)** for PR 18630 at commit [`c559a08`](https://github.com/apache/spark/commit/c559a08028633394371e784930dda0f8ff3f19ea).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79906/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -1315,6 +1294,80 @@ private[spark] object SparkSubmitUtils {
 }
 
+private [deploy] object DependencyUtils {
+  def resolveMavenDepndencies(args: Option[SparkSubmitArguments] = None): String = {
```
nit: add empty line before this one, remove the empty line after it.

It's also weird for this method to depend on `SparkSubmitArguments`, and fall back to system properties. It makes more sense for it to take all those argument explicitly, and let the caller defined where they come from.

Finally there is a typo in the method name.

skonto replied:
Ok sure will change it.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -1315,6 +1294,80 @@ private[spark] object SparkSubmitUtils {
+      "spark.ssl.fs.protocol", "spark.ssl.protocol")
+    var jars = jarList
+
+    securityProperties
```
This is really oddly indented.

```
securityProperties.map { pName => 
  sys.props.get(pName).map { pValue =>
    sparkProperties.put(pName, pValue)
  }
}
```

Also, it doesn't make a ton of sense to use `map` when you're not using the return value. That's what `foreach` is for.

skonto replied:
Yes correct.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -1315,6 +1294,80 @@ private[spark] object SparkSubmitUtils {
+
+  def addJarsToClassPath(jars: String, loader: MutableURLClassLoader): Unit = {
+    val childClasspath = new ArrayBuffer[String]()
+    if (jars != null) {childClasspath ++= jars.split(",")}
```
```
if (blah) {
  // code
}
```

But really you don't need `childClasspath` here at all.

skonto replied:
You mean simply split and add each jar? Yes should be more compact.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -1315,6 +1294,80 @@ private[spark] object SparkSubmitUtils {
 
 }
 
+private [deploy] object DependencyUtils {
```
Do you mind moving this to a separate source file?

I think there's other stuff in `SparkSubmit` that should eventually be moved here, but it's ok to leave that for later.

Also, nit: no space after `private`.

skonto replied:
ok

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +70,16 @@ object DriverWrapper {
+      jars = SparkSubmit.mergeFileLists(jars, resolvedMavenCoordinates)
+    }
+
+    jars = DependencyUtils.resolveAndDownloadJars(jars, userJar)
```
Instead of using a `var` I find it cleaner to define a different variable (e.g. `val localJars = DependencyUtils...`).

skonto replied:
Ok no problem.

vanzin reviewed (commented):
A few minor things otherwise looks ok.

skonto commented:
@vanzin fixed the issues. Please give it another try or merge.

SparkQA commented:
**[Test build #80208 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80208/testReport)** for PR 18630 at commit [`70649e2`](https://github.com/apache/spark/commit/70649e2a130daa01e4d0da28c69c96b1bc9f26a3).

SparkQA commented:
**[Test build #80208 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80208/testReport)** for PR 18630 at commit [`70649e2`](https://github.com/apache/spark/commit/70649e2a130daa01e4d0da28c69c96b1bc9f26a3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80208/
Test PASSed.

skonto commented:
@srowen  or @vanzin can I get a merge pls?

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,97 @@
+
+  def createTempDir(): File = {
+    val targetDir = Files.createTempDirectory("tmp").toFile
+    // scalastyle:off runtimeaddshutdownhook
```
minor: you need to add `scalastyle:on` after this block.

skonto replied:
:+1:

skonto replied:
Fixed it.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,97 @@
+      "spark.ssl.fs.protocol", "spark.ssl.protocol")
+    var jars = jarList
+
+    securityProperties.foreach{pName =>
```
nit: `.foreach { pName =>`

Also, I suggested a different, clearer indentation previously that makes this block easier to read.

skonto replied:
ok missed that.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +70,24 @@ object DriverWrapper {
   }
+
+  private def setupDependencies(loader: MutableURLClassLoader, userJar: String): Unit = {
+    val (packagesExclusions, packages, repositories, ivyRepoPath) = {
```
Slightly less verbose:

    val Seq(a, b, c, d) = Seq("a", "b", "c", "d").map(sys.props.get(_).orNull)

skonto replied:
ok sounds good.

skonto replied:
:+1:

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,97 @@
+    securityProperties.foreach{pName =>
+      sys.props.get(pName).foreach {pValue => sparkProperties.put(pName, pValue)}}
+
+    jars = Option(jars).map(SparkSubmit.resolveGlobPaths(_, hadoopConf)).orNull
```
Instead of re-assigning to this variable, a more functional approach is both cleaner and easier to read:

```
Option(jars)
   .map(SparkSubmit....)
   .flatMap(_.split...)
   .filterNot(...)
   .mkString(...)
```

skonto replied:
This will not work as flatMap is applied on the Option not the array of split values.

vanzin replied:
I meant less "use my code verbatim" and more "rewrite this for readability".

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,97 @@
+      SparkSubmitUtils.buildIvySettings(Option(repositories), Option(ivyRepoPath))
+    }
+
+    val resolvedMavenCoordinates = SparkSubmitUtils.resolveMavenCoordinates(
```
No need for the `val` since you're just returning it.

skonto replied:
Yeah ok its more for clarity but anyway sure.

skonto replied:
fixed it.

vanzin reviewed (commented):
Some style nits, otherwise looks ok.

skonto commented:
@vanzin fixed the issues. Ready for merge.

SparkQA commented:
**[Test build #80433 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80433/testReport)** for PR 18630 at commit [`a177992`](https://github.com/apache/spark/commit/a177992ec0a5a6797422c6b445c95b5b50d436b5).

SparkQA commented:
**[Test build #80433 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80433/testReport)** for PR 18630 at commit [`a177992`](https://github.com/apache/spark/commit/a177992ec0a5a6797422c6b445c95b5b50d436b5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80433/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,101 @@
+    val securityProperties = List("spark.ssl.fs.trustStore", "spark.ssl.trustStore",
+      "spark.ssl.fs.trustStorePassword", "spark.ssl.trustStorePassword",
+      "spark.ssl.fs.protocol", "spark.ssl.protocol")
+    var jars = jarList
```
nit: not necessary anymore

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,101 @@
+          .mkString(",")
+      }
+      .filterNot(_ == "")
+      .map (SparkSubmit.downloadFileList(_, targetDir, sparkProperties, hadoopConf))
```
nit: no space after `map`

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,101 @@
+      "spark.ssl.fs.protocol", "spark.ssl.protocol")
+    var jars = jarList
+
+    securityProperties.map { pName =>
```
`map` -> `foreach`.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,101 @@
+    var jars = jarList
+
+    securityProperties.map { pName =>
+      sys.props.get(pName).map { pValue =>
```
`map` -> `foreach`.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +70,24 @@ object DriverWrapper {
   }
+
+  private def setupDependencies(loader: MutableURLClassLoader, userJar: String): Unit = {
+
```
nit: remove blank line.

BryanCutler commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -300,28 +298,14 @@ object SparkSubmit extends CommandLineUtils {
-        ivySettings, exclusions = exclusions)
-
+      val resolvedMavenCoordinates =
+      DependencyUtils.resolveMavenDepndencies(args.packagesExclusions, args.packages,
```
should be an indention here.  Maybe this instead?
```
val resolvedMavenCoordinates = DependencyUtils.resolveMavenDepndencies(args.packagesExclusions, 
  args.packages, ivySettings, exclusions = exclusions)
```

BryanCutler commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,101 @@
+      }
+      .filterNot(_ == "")
+      .map (SparkSubmit.downloadFileList(_, targetDir, sparkProperties, hadoopConf))
+      .orNull
```
This is a huge statement.  I think it would be better to break it up to be easier to follow.

vanzin replied:
What's your suggestion?

This syntax is more functional and easier to follow than a bunch of disjoint statements.

BryanCutler replied:
Here is a suggestion, not much different but I prefer some separation between the calls to `resolveGlobPaths` and `downloadFileList` since there is a lot going on in each of those, and checking `nonEmpty` vs ` .filterNot(_ == "")`

```scala
def resolveAndDownloadJars(jars: String, userJar: String): String = {
   ...
    Option(jars).map { allJars =>
      val jarList = SparkSubmit.resolveGlobPaths(allJars, hadoopConf)
        .split(",")
        .filterNot(_.contains(userJar.split("/").last))

      if (jarList.nonEmpty) {
        SparkSubmit.downloadFileList(jarList.mkString(","), targetDir, sparkProperties, hadoopConf)
      } else {
        null
      }
    }.orNull
  }
```
it's up to you though, not a huge deal.  If you do leave it as is, the `var jar` should be removed.

BryanCutler commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,101 @@
+
+import org.apache.spark.util.MutableURLClassLoader
+
+private[deploy] object DependencyUtils {
```
Do we really need this new object?  It make the order of calls confusing.  For instance SparkSubmit calls DependencyUtils which makes calls to SparkSubmitUtils.  Can we just move this to SparkSubmitUtils?

vanzin replied:
I asked for this change, because `SparkSubmitUtils` is kind of a bad place to keep this code if things other than `SparkSubmit` start using it.

BryanCutler replied:
Ok, that's fine if you are thinking the calls back to `SparkSubmitUtils` can be moved here eventually.

BryanCutler commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -66,4 +70,24 @@ object DriverWrapper {
+    val jars = {
+      val jarsProp = sys.props.get("spark.jars").orNull
+      if (!StringUtils.isBlank(resolvedMavenCoordinates)) {
+        SparkSubmit.mergeFileLists(jarsProp, resolvedMavenCoordinates)
```
is this same code repeated elsewhere?  if so maybe it could be put into a common function

BryanCutler reviewed (changes requested):
Overall this looks fine to me.  I would prefer just keeping the common code in `SparkSubmitUtils` instead of adding a new object.  Is there no tests that can be added to test this behavior?  Also, I think some of the logic in SparkSubmit works to provide py files along with the jars.  If that is also not working for standalone cluster, maybe it can easily be added with this change too?

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,101 @@
+    SparkSubmitUtils.resolveMavenCoordinates(packages, ivySettings, exclusions = exclusions)
+  }
+
+  def createTempDir(): File = {
```
This code cannot call into things that initialize logging, and the Utils method initializes logging indirectly through `ShutdownHookManager`.

BryanCutler replied:
Ok, makes sense thanks for clarifying that

skonto commented:
@vanzin I fixed these minor issues.

SparkQA commented:
**[Test build #80468 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80468/testReport)** for PR 18630 at commit [`c0b0a7d`](https://github.com/apache/spark/commit/c0b0a7d79ca27bbcf91245b3d80070d5d4188174).

skonto commented:
@BryanCutler @vanzin  to make things testable DriverWrapper needs refactoring from a quick look I took. 
 py files are resolved in client mode, let's fix it in another PR (I could do it). The docs (https://spark.apache.org/docs/latest/submitting-applications.html) state:
"Currently, standalone mode does not support cluster mode for Python applications."
So is the file distribution the only thing to do? I havent scoped the work needed to support python apps.

vanzin commented:
I wasn't really expecting python support to be added here. I wonder if there's a bug open for that.

BryanCutler commented:
Sure, python support could be added at a later point, I was just thinking if it was only a small addition to what's already here, but no problem.  Btw, after checking out this PR I tried spark-shell and got the error below.  Not sure if it was my environment, but after switching back to master it worked fine
```
bin/spark-shell 
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/commons/logging/LogFactory
	at org.apache.hadoop.conf.Configuration.<clinit>(Configuration.java:178)
	at org.apache.spark.deploy.SparkSubmit$.prepareSubmitEnvironment(SparkSubmit.scala:324)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:155)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:121)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.ClassNotFoundException: org.apache.commons.logging.LogFactory
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
```

skonto commented:
@BryanCutler you just started spark shell and it failed? How can I reproduce it?

BryanCutler commented:
Yeah, just by running `bin/spark-shell` it failed immediately with that error.  I double-check by rebuilding and same thing but I'm not sure if was something from your changes or not.  Are you able to startup the shell?

skonto commented:
> spark-2.3.0-SNAPSHOT-bin-18630/bin$ ./spark-shell --verbose --master spark://ip-10-10-1-79:7077 
Using properties file: null
Parsed arguments:
  master                  spark://ip-10-10-1-79:7077
  deployMode              null
  executorMemory          null
  executorCores           null
  totalExecutorCores      null
  propertiesFile          null
  driverMemory            null
  driverCores             null
  driverExtraClassPath    null
  driverExtraLibraryPath  null
  driverExtraJavaOptions  null
  supervise               false
  queue                   null
  numExecutors            null
  files                   null
  pyFiles                 null
  archives                null
  mainClass               org.apache.spark.repl.Main
  primaryResource         spark-shell
  name                    Spark shell
  childArgs               []
  jars                    null
  packages                null
  packagesExclusions      null
  repositories            null
  verbose                 true

Spark properties used, including those specified through
 --conf and those from the properties file null:
  

    
Main class:
org.apache.spark.repl.Main
Arguments:

System properties:
(SPARK_SUBMIT,true)
(spark.app.name,Spark shell)
(spark.jars,)
(spark.submit.deployMode,client)
(spark.master,spark://ip-10-10-1-79:7077)
Classpath elements:



Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
17/08/09 23:28:03 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://10.10.1.79:4040
Spark context available as 'sc' (master = spark://ip-10-10-1-79:7077, app id = app-20170809232804-0003).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 2.3.0-SNAPSHOT
      /_/
         
Using Scala version 2.11.8 (OpenJDK 64-Bit Server VM, Java 1.8.0_131)
Type in expressions to have them evaluated.
Type :help for more information.

scala>

BryanCutler commented:
Maybe it was just something with my env - but I was running it locally, can you just verify that works too?  Just don't specify the `--master` conf and run out of your spark home dir

skonto commented:
@BryanCutler sure check here, it works:
https://gist.github.com/skonto/dc2070d1529c97ec5de32e99983a834f

skonto commented:
This is how I build things:

./build/mvn -Pmesos -Phadoop-2.7 -Dhadoop.version=2.7.0 -DskipTests clean package

#  -DskipTests clean package

export JAVA_HOME=/usr/lib/jvm/java-8-oracle/jre/

./dev/make-distribution.sh --name 18630 --tgz -Phadoop-2.7 -Pmesos

BryanCutler commented:
Ok, thanks for checking.  It doesn't look like it's coming from your changes, so I'm sure it's just me.

SparkQA commented:
**[Test build #80468 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80468/testReport)** for PR 18630 at commit [`c0b0a7d`](https://github.com/apache/spark/commit/c0b0a7d79ca27bbcf91245b3d80070d5d4188174).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80468/
Test PASSed.

skonto commented:
@vanzin do you think we are ready for a merge?

vanzin commented on core/src/main/scala/org/apache/spark/deploy/DependencyUtils.scala:
```diff
@@ -0,0 +1,100 @@
+
+private[deploy] object DependencyUtils {
+
+  def resolveMavenDepndencies(
```
Your method name still has a typo.

skonto replied:
ok

skonto commented:
@vanzin fixed.

SparkQA commented:
**[Test build #80549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80549/testReport)** for PR 18630 at commit [`db60b27`](https://github.com/apache/spark/commit/db60b273e971dc758c5ff09ca3660f7f63522392).

SparkQA commented:
**[Test build #80549 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80549/testReport)** for PR 18630 at commit [`db60b27`](https://github.com/apache/spark/commit/db60b273e971dc758c5ff09ca3660f7f63522392).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80549/
Test PASSed.

vanzin commented:
You forgot to address @BryanCutler 's comments; I'll fix the easy ones during merge.

Merging to master.

skonto commented:
@vanzin I dont forgot I didnt see agreement.
